### PR TITLE
MRG: Make y a required parameter in CSP.fit_transform()

### LINKED ISSUE #10079: CSP.fit() requires y to work, however the parameter is optional in fit_transform()
Reported on the forum:

https://mne.discourse.group/t/can-not-do-csp-fit-transform-without-label/4072

Calling `CSP.fit_transform()` without `y` (which is alright according to the API, which defaults `y=None`) yields the following exception:

```python
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/long/anaconda/envs/3.6/lib/python3.6/site-packages/mne/decoding/mixin.py", line 30, in fit_transform
    return self.fit(X, **fit_params).transform(X)
TypeError: fit() missing 1 required positional argument: 'y'
```

This is because the `fit_transform()` method inherited from `TransformerMixin` is not compatible with the `fit` method implemented in `CSP`.

Solution is to give `CSP` its very own `fit_transform` method that doesn't make `y` an optional parameter.

### PR DESCRIPTION
**Pull request by hoechenberger**
Fixes #10079

### CODE REVIEW

File: mne/decoding/csp.py [hunk 1]
```diff
@@ -231,6 +231,10 @@ def transform(self, X):
                 X /= self.std_
         return X
 
+    @copy_doc(TransformerMixin.fit_transform)
+    def fit_transform(self, X, y, **fit_params):
+        return super().fit_transform(X, y=y, **fit_params)
+
```
Comment (larsoner) [inline mne/decoding/csp.py:237]: Can't we just change `fit(X, y=None)` instead?

Comment (larsoner) [reply mne/decoding/csp.py:237]: Never mind, I saw your reply in #10079

**Comment by hoechenberger** [comment]
@larsoner Any idea how to fix the docstring test?

**Comment by larsoner** [comment]
I would add a docstring manually. Maybe you can deduplicate and make use of fill_doc?

**Comment by hoechenberger** [comment]
> I would add a docstring manually.

The thing is, I'm using `copy_doc`, I thought this should take care of such things 🤔

File: mne/decoding/csp.py [hunk 2]
```diff
@@ -231,6 +231,10 @@ def transform(self, X):
                 X /= self.std_
         return X
 
+    @copy_doc(TransformerMixin.fit_transform)
+    def fit_transform(self, X, y, **fit_params):
```
Comment (larsoner) [inline mne/decoding/csp.py:235]: Oh it's just pydocstyle, not numpydoc -- if you `git grep copy_doc` to see where it's used elsewhere, e.g.:

https://github.com/mne-tools/mne-python/blob/8954c3711b57dc1b90a8079c278384ba684ed17f/mne/io/base.py#L1029-L1035

And cross-ref to the error:
```
        D102: Missing docstring in public method
```

You can see that this in principle should fix it:
```suggestion
    def fit_transform(self, X, y, **fit_params):  # noqa: D102
```

**Comment by hoechenberger** [comment]
Thanks, Eric! This seems to do the trick

**Review by agramfort — APPROVED** [review]

**Comment by larsoner** [comment]
Worth a `latest.inc` or no?

**Comment by hoechenberger** [comment]
It doesn't really change the behavior – code that doesn't pass `y` now simply fails earlier. Hence, I don't think we'll need a changelog entry here

**Comment by agramfort** [comment]
+1

**Comment by larsoner** [comment]
Thanks @hoechenberger !

**Pull request merged**